pkg/server/public/handlers/foo/create.go

```diff
@@ -0,0 +1,61 @@
+package foo
+
+import (
+	"net/http"
+
+	"github.com/emicklei/go-restful/v3"
+	"github.com/nrc-no/core/pkg/api/meta"
+	"github.com/nrc-no/core/pkg/api/types"
+	"github.com/nrc-no/core/pkg/api/types/validation"
+	"github.com/nrc-no/core/pkg/logging"
+	"github.com/nrc-no/core/pkg/utils"
+	uuid "github.com/satori/go.uuid"
+	"go.uber.org/zap"
+)
+
+func (h *Handler) Create() http.HandlerFunc {
+	return func(w http.ResponseWriter, req *http.Request) {
+		ctx := req.Context()
+		l := logging.NewLogger(ctx)
+
+		var reads types.FooReads
+		if err := utils.BindJSON(req, &reads); err != nil {
+			l.Error("Failed to bind request body", zap.Error(err))
+			utils.ErrorResponse(w, err)
+			return
+		}
+
+		if errList := validation.ValidateFooReads(&reads); !errList.IsEmpty() {
+			err := meta.NewInvalid(types.FooGR, "", errList)
+			l.Error("Failed to validate request body", zap.Error(err))
+			utils.ErrorResponse(w, err)
+			return
+		}
+
+		// I don't like how there's no validation that each field is set.
```

Comment:
I usually name the fields with the fact that the zero value is the default, like
`enableDangerousLogs (default = false)`, `disableTls (default = false)`

If you really need the user to provide a value, you could set up a kind of enum (with `iota` + go genereate & `stringer` package) so that you enforce a value there.

But you need to keep in mind that this does not just apply to deserialization. You can pass any struct to any method and not put in some fields. It's the way it was designed.. 

Some people create mandatory constructors + do not expose the structs outside of the package sometimes. 

I think that if the validation package is well tested, and it catches the errors, then there is no real concern here..

Reply:
> But you need to keep in mind that this does not just apply to deserialization
Agreed, but I found it easier to cause some significant issues when it's not validated for deserialisation.

> Some people create mandatory constructors + do not expose the structs outside of the package sometimes.
Yes, I was thinking of doing something like this. I may give it a go.

Reply:
Although actually I'm not sure how the constructor pattern would work:
```
# types/foo.go
package types

type foo struct {
  ...
}

func NewFoo( ... ) *foo {
  return &foo{ ... }
}
```

```
# store/foo.go

import "github.com/nrc-no/core/pkg/api/types"

# How do we get types.foo without exposing the struct?
func (d *fooStore) Create(ctx context.Context, foo *types.foo) (*types.foo, error) {
  ...
}
```

Reply:
You can't pass non-exported types across packages...

Reply:
Yeah exactly, so not sure this really works

Reply:
I think it's too much investment for what it gives back... One thing you could do is to implement the `UnmarshalJSON` interface on your type, and verify that the types are there.
```
func (m *MyType) UnmarshalJSON(b []byte) error {
temp := map[string]interface{}
if err := json.Unmarshal(b, &temp); err != nil {
return err
}
if _, ok := temp["myJsonField"]; !ok {
  return fmt.ErrFmt("missing myJsonField")
}
return json.Marshal(b, &m)

}
```